http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java

```diff
@@ -1070,30 +1075,26 @@ public synchronized void onHeaders(Response response)
             if (length > maxLength) {
                 response.abort(new ResponseTooLargeException());
             }
-            if (length > buffer.length) {
-                buffer = Arrays.copyOf(buffer, Ints.saturatedCast(length));
-            }
+            allocateCurrentArray();
         }
 
         @Override
         public synchronized void onContent(Response response, ByteBuffer content)
         {
             int length = content.remaining();
-            int requiredCapacity = size + length;
-            if (requiredCapacity > buffer.length) {
-                if (requiredCapacity > maxLength) {
-                    response.abort(new ResponseTooLargeException());
-                    return;
-                }
-
-                // newCapacity = min(log2ceiling(requiredCapacity), maxLength);
-                int newCapacity = min(Integer.highestOneBit(requiredCapacity) << 1, maxLength);
-
-                buffer = Arrays.copyOf(buffer, newCapacity);
+            size += length;
+            if (size > maxLength) {
+                response.abort(new ResponseTooLargeException());
+                return;
             }
 
-            content.get(buffer, size, length);
-            size += length;
+            while (length > 0) {
+                allocateCurrentArray();
```

Comment:
In this case I'd prefer to see
```java
if (currentBufferPosition >= currentBuffer.length) {
     allocateNewBuffer();
}
```
This leaves the check in the main loop and puts only the allocation (i.e., the slow path) in a separate method.  Of course you should validate in the allocate method that there is not remaining space and throw and exception if not.